mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,343 @@
+    rx: Arc<AsyncMutex<UnboundedReceiver<ControllerMessage>>>,
+}
+
+enum ControllerMessage {
+    Request(WsRequest),
+    Close,
+}
+
+#[derive(Debug)]
+struct WsRequest {
+    serialized_request: String,
+    request_id: RequestId,
+    response_notifier: oneshot::Sender<()>,
+}
+
+/// A wrapper type for raw pointers used as a mutable Send & Sync HashMap reference.
+///
+/// Safety notes:
+///
+/// The implemented algorithm for socket request-response is already thread-safe,
+/// so we don't care about race conditions.
+///
+/// As for deallocations, see the `Drop` implementation below.
+#[derive(Debug)]
+struct SafeMapPtr(*mut HashMap<RequestId, Vec<u8>>);
+
+impl Drop for SafeMapPtr {
+    fn drop(&mut self) {
+        // Let the compiler do the job.
+        let _ = unsafe { Box::from_raw(self.0) };
+    }
+}
+
+unsafe impl Send for SafeMapPtr {}
+unsafe impl Sync for SafeMapPtr {}
+
+impl WebsocketTransport {
+    pub(crate) fn with_event_handlers(
+        node: WebsocketTransportNode,
+        event_handlers: Vec<RpcTransportEventHandlerShared>,
+    ) -> Self {
+        let (req_tx, req_rx) = futures::channel::mpsc::unbounded();
+
+        WebsocketTransport {
+            node,
+            event_handlers,
+            responses: Arc::new(SafeMapPtr(Box::into_raw(Default::default()))),
+            request_id: Arc::new(AtomicUsize::new(1)),
+            controller_channel: ControllerChannel {
+                tx: Arc::new(AsyncMutex::new(req_tx)),
+                rx: Arc::new(AsyncMutex::new(req_rx)),
+            }
+            .into(),
+            connection_guard: Arc::new(AsyncMutex::new(())),
+            gui_auth_validation_generator: None,
+            last_request_failed: Arc::new(AtomicBool::new(false)),
+        }
+    }
+
+    pub(crate) async fn start_connection_loop(self) {
+        let _guard = self.connection_guard.lock().await;
+
+        const MAX_ATTEMPTS: u32 = 3;
+        const SLEEP_DURATION: f64 = 1.;
+        const SIMPLE_REQUEST: &str = r#"{"jsonrpc":"2.0","method":"net_version","params":[],"id": 0 }"#;
+        const KEEPALIVE_DURATION: Duration = Duration::from_secs(10);
+
+        async fn attempt_to_establish_socket_connection(
+            address: String,
+            max_attempts: u32,
+            sleep_duration_on_failure: f64,
+        ) -> tokio_tungstenite_wasm::Result<WebSocketStream> {
+            loop {
+                let mut attempts = 0;
+
+                match tokio_tungstenite_wasm::connect(address.clone()).await {
+                    Ok(ws) => return Ok(ws),
+                    Err(e) => {
+                        attempts += 1;
+                        if attempts > max_attempts {
+                            return Err(e);
+                        }
+
+                        Timer::sleep(sleep_duration_on_failure).await;
+                        continue;
+                    },
+                };
+            }
+        }
```

Comment:
review note: you should move `let mut attempts = 0;` from the loop.

suggestion:
Lets use exponential backoff strategy? which increases the delay between consecutive retry attempts in an exponential manner. 

we can have something like this
```rust
pub(crate) async fn start_connection_loop(self) {
    let _guard = self.connection_guard.lock().await;

    const MAX_ATTEMPTS: u32 = 3;
    const INITIAL_SLEEP_DURATION: f64 = 1.0;
    const KEEPALIVE_DURATION: Duration = Duration::from_secs(10);
    const MAX_SLEEP_DURATION: f64 = 32.0;

    async fn attempt_to_establish_socket_connection(
        address: String,
        max_attempts: u32,
    ) -> tokio_tungstenite_wasm::Result<WebSocketStream> {
        let mut attempts = 0;
        let mut sleep_duration = INITIAL_SLEEP_DURATION;

        loop {
            match tokio_tungstenite_wasm::connect(address.clone()).await {
                Ok(ws) => return Ok(ws),
                Err(e) => {
                    attempts += 1;
                    if attempts > max_attempts {
                        return Err(e);
                    }

                    Timer::sleep(Duration::from_secs_f64(sleep_duration)).await;
                    // Double the wait time but no more than MAX_SLEEP_DURATION
                    sleep_duration = (sleep_duration * 2.0).min(MAX_SLEEP_DURATION); 
                },
            };
        }
    }
}
```

Such approach can reduce the load on server. as we increase the wait time between retries, it gives server more time to recover in case it is having high load or other temporary issues.
Also this strategy is kinda flexible, as we can tweak the initial delay, maximum delay, and the maximum number of attempts based on the expected behavior of the server and network conditions.

Reply:
Done